src/main/java/edu/byu/cs/service/UserService.java

```diff
@@ -0,0 +1,139 @@
+        if (repoUrl == null && netId == null) {
+            throw new BadRequestException("You must provide either a repoUrl or a netId");
+        }
+
+        try {
+            if (repoUrl != null) {
+                updates.addAll(DaoService.getRepoUpdateDao().getUpdatesForRepo(repoUrl));
+            }
+            if (netId != null) {
+                updates.addAll(DaoService.getRepoUpdateDao().getUpdatesForUser(netId));
+            }
+        } catch (Exception e) {
+            LOGGER.error("Error getting repo updates:", e);
+            throw new InternalServerException(e);
+        }
+
+        return updates;
+    }
+
+    private static void setRepoUrl(String studentNetId, String repoUrl, String adminNetId) throws BadRequestException, InternalServerException, PriorRepoClaimBlockageException {
+        try {
+            if (!isValidRepoUrl(repoUrl)) {
+                throw new BadRequestException("Invalid Github Repo Url. Check if the link is valid and points directly to a Github Repo.");
+            }
+        } catch (BadRequestException e) {
+            throw e;
+        } catch (Exception e) {
+            LOGGER.error("Error cloning repo during repoPatch: {}", e.getMessage());
+            throw new InternalServerException("There was an internal server error in verifying the Github Repo", e);
+        }
+
+        RepoUpdate historicalUpdate;
+        try {
+            historicalUpdate = verifyRepoIsAvailableForUser(repoUrl, studentNetId);
+        } catch (DataAccessException e) {
+            throw new InternalServerException("There was an internal server error in checking the repo update history for this url", e);
+        }
+        if (historicalUpdate != null) {
+            if (adminNetId != null) {
+                throw new PriorRepoClaimBlockageException("Repo is blocked because of a prior claim: " + historicalUpdate);
+            } else {
+                LOGGER.info("Student {} was blocked from updating their url because of a prior claim: {}", studentNetId, historicalUpdate);
+                throw new PriorRepoClaimBlockageException("Please talk to a TA to submit this url");
+            }
+        }
+
+        try {
+            RepoUpdate update = new RepoUpdate(Instant.now(), studentNetId, repoUrl, adminNetId != null, adminNetId);
+            DaoService.getUserDao().setRepoUrl(studentNetId, repoUrl);
+            DaoService.getRepoUpdateDao().insertUpdate(update);
+        } catch (DataAccessException e) {
+            throw new InternalServerException("There was an internal server error in saving the GitHub Repo URL", e);
+        }
+
+        if (adminNetId == null) {
+            LOGGER.info("student {} changed their repoUrl to {}", studentNetId, repoUrl);
+        } else {
+            LOGGER.info("admin {} changed the repoUrl for student {} to {}", adminNetId, studentNetId, repoUrl);
+        }
+    }
+
+    private static boolean isValidRepoUrl(String url) {
```

Comment:
I found the code that performs this work in `GitHelper`! We may have to expose the private function, but let's prefer to reuse code rather than allowing it to be duplicated.

Reply:
I agree it shouldn't be here. That would introduce a secondary entry point and usage context for GitHelper, though, and at present GitHelper is very nicely self-contained and cohesive. Rather than broaden the Single Responsibility of GitHelper, I propose we rename GitHelper to GitGradingHelper or something similar, then create a GitHelper class containing pure static functions that don't depend on a GradingContext. Then both UserService and GitGradingHelper can depend on it.

Objections?

Reply:
Moving conversation to #463